Simulation Structure - Health Modeling (#12)
SOMAS2021/SOMAS-Final-Report: 002_simulation_structure/002_simulation_structure.tex

\chapter{Simulation Structure}\label{simulation_structure}

%%%%%%%%
%%%%%%%% Simulation Environment
%%%%%%%%
\section{Simulation Environment}\label{simulation_environment}
Goes over the backend design

%%%%%%%%
%%%%%%%% Simulation Flow
%%%%%%%%
\section{Simulation Flow}\label{simulation_flow}
The order in which events happen
Need a nice diagram for this

%%%%%%%%
%%%%%%%% Message Passing
%%%%%%%%
\section{Message Passing}\label{message_passing}
Need a nice diagram for this

%%%%%%%%
%%%%%%%% Health Modeling
%%%%%%%%
\section{Health Modeling}\label{health_modeling}

\ToDo{Add diagram with a possible scenario of several days in a row?}

%
%%%%%%%% Global Description
%
\subsection{Global Description}

The health of the agents living in the tower is represented by Health Points (HP). Two mechanisms affect an agent's HP: how much food they eat, and their ``cost of living''. The cost of living represents how many calories a human needs to eat each day to stay healthy. These two mechanisms are implemented using the functions \lstinline$updateHP$ and \lstinline$hpDecay$, respectively. These two functions are described below.

> At the end of each day, agents are assigned an HP value based on how much food they have eaten and their cost of living. This HP value is an integer and has a maximum value of \lstinline$MaxHP$, and a minimum value of \lstinline$HPCritical$. As its name suggests, \lstinline$HPCritical$ is a critical HP value for the agents: they can only survive a certain number of days (\lstinline$MaxDayCritical$) at this level. When in the critical state, if agents can increase their HP by \lstinline$HPReqCToW$ (``HP Required to move from Critical To Weak''), then they move into the ``weak state'' (\Cref{fig:health_system}), and their HP takes the value of \lstinline$WeakLevel$. The amount that an agent's HP increases from eating is determined by the function \lstinline$updateHP()$.

Reviewer: Ideally, you refer to the fig:health_system diagram in the text.

\begin{figure}[htb]
    \centering

>     \includegraphics[width=0.3\linewidth]{002_simulation_structure/images/health_global.pdf}

Reviewer: Is there a reason a zoomed in version of the diagram is shown instead of the full one? At the moment it seems like this doesn't need to be a diagram and it could just be a list of health levels.

Author: The diagram is not a zoomed version, as we have an additional axis on this one: FoodTaken. The idea is to relate the curve on the right with the general representation of HP as a function of FoodTaken

Reviewer: The image looks like its a graph with the other axis cropped out. That may just be my github not liking the uploaded image though

Author: ![image](https://user-images.githubusercontent.com/80003671/148142275-19e43334-ca78-4bf2-9b3f-43f4dd65c542.png)

Author: It should be something like that. Is it what you also see? @Silvertrousers 
EDIT: nvm, you are referring to the Figure 3.1

Reviewer: huh, I see this: 
![image](https://user-images.githubusercontent.com/56201218/148142441-34237fb5-53dd-475c-942a-b4713142a102.png)

Reviewer: for the health_global.pdf file

Author: I read your comment incorrectly. You are referring to Fig 3.1, which is indeed rather simple. We could have the special levels of the graph (WeakLevel, MaxHP,..) explained in the text only, but I think it is always helpful for the reader to have some diagrams. I think it is always good to have a bit too many diagrams too not enough, especially if we don't have a limited amount of space. In this case, we first present the special levels (Fig. 3.1), and then show the updateHP function (Fig. 3.2) with the idea that the reader already know a part of the diagram, so that it is easier to read it.
What I would propose is to leave it there for now, and delete it if needed when the state of the report is more advanced. In all cases, there are some red comments that will be addressed in the coming days. What do you think about this @Silvertrousers ?

Reviewer: alright, lets leave it in for now

    \caption{The health of the agents is represented by a HP value between \lstinline$HPCritical$ and \lstinline$MaxHP$. All HP values which are below \lstinline$WeakLevel$ are classed as critical. The diagram is not drawn to scale.}
    \label{fig:health_system}
\end{figure}

\subsection{Food and Health: \lstinline$updateHP$}\label{updateHP}

> To increase their HP, agents need to eat. However, the amount an agent's HP improves can saturate in a single day; eating more than a certain amount will provide an agent with no extra benefit to their HP. Moreover, eating more food will lead to diminishing returns in terms of HP change. Mathematically, the ideas of diminishing returns and saturation are well captured by the step response of a 1st-order system:

Reviewer: should refer to the equation in the text so change the last line to: "Mathematically, the ideas of diminishing returns and saturation are well captured by the step response of a 1st-order system shown by Equation \ref{updateHP_general}."

> To increase their HP, agents need to eat. However, the amount an agent's HP improves can saturate in a single day; eating more than a certain amount will provide an agent with no extra benefit to their HP. Moreover, eating more food will lead to diminishing returns in terms of HP change. Mathematically, the ideas of diminishing returns and saturation are well captured by the step response of a 1st-order system:

Reviewer: Should refer to the equation in the text using (equation_number) so change the last line to: "Mathematically, the ideas of diminishing returns and saturation are well captured by the step response of a 1st-order system, (\ref{updateHP_general})."

\begin{equation}\label{updateHP_general}
   \text{newHP}= \text{currentHP} +\underbrace{w(1-e^{\frac{-\text{foodTaken}}{\tau}})}_{\text{HPChange}}
\end{equation}

> where the two parameters $w$ and $\tau$ are defined at the beginning of the simulation. The shape of this curve is given in \Cref{fig:updateHP} together with some important parameters.

Reviewer: If you make the above change, the start of this line can become: "The two parameters ... " (removing the where)

Author: done

\begin{figure}[htb]%
    \centering
    \subfloat[\centering Overview]{{\includegraphics[width=0.36\linewidth]{002_simulation_structure/images/health_updateHP_overview.pdf}}}%
    \qquad
    \subfloat[\centering Detailed representation]{{\includegraphics[width=0.36\linewidth]{002_simulation_structure/images/health_updateHP_detailed.pdf}}}%
    \caption{\texttt{updateHP} as a function of the amount of food eaten (``FoodTaken'').}%
    \label{fig:updateHP}%
\end{figure}

It is not possible to gain more HP than $w$ over the duration of one day; this is an intentional limit to prevent an agent's health from improving too quickly. As an example, we can think of an agent that starts from the weak level and wants to reach the maximum HP value. It would take several days for this agent to ``recover'' from this weak level and stabilise its health to a high HP value. 

Note that it is possible for an agent to achieve an HP value that is larger than \texttt{MaxHP} inside \lstinline$hpDecay()$. At the end of each day, the \lstinline$hpDecay()$ function will apply the cost of living and then bound the final HP value by \lstinline$MaxHP$.

> Agents in the critical state are treated differently. For these agents, HP is updated according to this equation:

Reviewer: again, ideally, you refer to the equation in the text body: "HP is updated according to (\ref{updateHP_critical})"

\begin{equation}\label{updateHP_critical}
    \text{newHP} = \min\left\{\text{HPCritical}+\text{HPReqCToW}, \text{currentHP} +w(1-e^{\frac{-\text{foodTaken}}{\tau}})\right\}
\end{equation}

\subsection{Cost of Living: \lstinline$hpDecay()$}\label{hpDecay}

> At the end of each day, the HP value of the agents will be reduced by the cost of living. The cost of living is larger for an agent with larger HP value than for an agent with lower HP value. This fact is motivated by a simple observation: humans that have stronger bodies and immune systems also need more food to sustain their level of health. The exact relation between HP value, cost of living, and HP value after applying the cost of living is given by the following linear relation:

Reviewer: see above comments about referring to equations in text. Change last line to: "The exact relation between HP value, cost of living, and HP value after applying the cost of living is given by the linear relation (\ref{hpDecay_equation})"

Author: done

\begin{equation}\label{hpDecay_equation}
    \text{newHP} = \text{currentHP}-\left[b + s(\text{currentHP}-\text{WeakLevel})\right]
\end{equation}

> where $b$ is a (constant) base cost, and $s$ is the slope of the linear function. These parameters are initialised at the beginning of the simulation. \ToDo{Add a diagram for this?}

Reviewer: see above about referring to equation parameters. Change to "$b$ ...." removing the where

Author: done

> To ensure that the HP value at the end of the day is bounded by \texttt{MaxHP}, we slightly modify the above expression:

Reviewer: "we slightly modify (\ref{hpDecay_equation}) to produce (\ref{hpDecay_bounded})"

Author: done

> 

Reviewer: I think the following description could be explained with regular text and a state diagram of some sort. The equations don't really say much here.

Author: I think it is actually ok to have it for completeness. But I agree, this is not the most important equation

Reviewer: yeah thats fair, if we keep then we should at least keep a consistent reference system to the equations, i.e. refer to equations in text using(eqn_number) instead of by position.

\chapter{Simulation Structure}\label{simulation_structure}

%%%%%%%%
%%%%%%%% Simulation Environment
%%%%%%%%
\section{Simulation Environment}\label{simulation_environment}
Goes over the backend design

%%%%%%%%
%%%%%%%% Simulation Flow
%%%%%%%%
\section{Simulation Flow}\label{simulation_flow}
The order in which events happen
Need a nice diagram for this

%%%%%%%%
%%%%%%%% Message Passing
%%%%%%%%
\section{Message Passing}\label{message_passing}
Need a nice diagram for this

%%%%%%%%
%%%%%%%% Health Modeling
%%%%%%%%
\section{Health Modeling}\label{health_modeling}

\ToDo{Add diagram with a possible scenario of several days in a row?}

%
%%%%%%%% Global Description
%
\subsection{Global Description}

> The health of the agents living in the tower is represented by Health Points (HP). Two mechanisms affect the agents HP: the food they eat, and their cost of living. The costs of living can be compared to an amount of calories a human body is using to stay alive. These two mechanisms are implemented using the functions \texttt{updateHP} and \texttt{hpDecay}, respectively. These two functions are described in details below.

Reviewer 2: ```suggestion
The health of the agents living in the tower is represented by Health Points (HP). Two mechanisms affect an agent's HP: how much food they eat, and their ``cost of living''. The cost of living represents how many calories a human needs to eat each day to stay healthy. These two mechanisms are implemented using the functions \lstinline$updateHP$ and \lstinline$hpDecay$, respectively. These two functions are described below.
```

> At the end of every day, the agents are assigned an HP value based on the food they ate and their cost of living. This final HP value is an integer and can have a maximal value of \texttt{MaxHP}, and a minimal value of \texttt{HPCritical}. As its name suggests, \texttt{HPCritical} is a critical HP value for the agents: they can only survive a specific amount of days (named \texttt{MaxDayCritical}) at this level. Fortunately for the agents, they can upgrade their condition by going to the so-called "weak level" (\Cref{fig:health_system}). This level is defined as the HP value \texttt{WeakLevel} and every HP values below are described as "critical state". This upgrade happens if the agents are able to increase their HP of \texttt{HPReqCToW} (standing for "HP Required to go from Critical To Weak"). The way agents can increase their HP by eating is expressed by the function \texttt{updateHP}.

Reviewer 2: ```suggestion
At the end of each day, agents are assigned an HP value based on how much food they have eaten and their cost of living. This HP value is an integer and has a maximum value of \lstinline$MaxHP$, and a minimum value of \lstinline$HPCritical$. As its name suggests, \lstinline$HPCritical$ is a critical HP value for the agents: they can only survive a certain number of days (\lstinline$MaxDayCritical$) at this level. When in the critical state, if agents can increase their HP by \lstinline$HPReqCToW$ (``HP Required to move from Critical To Weak''), then they move into the ``weak state'' (\Cref{fig:health_system}), and their HP takes the value of \lstinline$WeakLevel$. The amount that an agent's HP increases from eating is determined by the function \lstinline$updateHP()$.
```

\begin{figure}[htb]
    \centering
    \includegraphics[width=0.3\linewidth]{002_simulation_structure/images/health_global.pdf}

>     \caption{The health of the agents is represented by a HP value between \texttt{HPCritical} and \texttt{MaxHP}. All HP values which are below \texttt{WeakLevel} are included in the critical state. The diagram is not drawn to scale.}

Reviewer 2: ```suggestion
    \caption{The health of the agents is represented by a HP value between \lstinline$HPCritical$ and \lstinline$MaxHP$. All HP values which are below \lstinline$WeakLevel$ are classed as critical. The diagram is not drawn to scale.}
```

    \label{fig:health_system}
\end{figure}

> %
> %%%%%%%% Food and Health: updateHP
> %
> \subsection{Food and Health: \texttt{updateHP}}\label{updateHP}
> To upgrade their HP, the agents need to eat. However, a certain saturation in the HP gain occurs when the agents keep eating. Moreover, we consider diminishing returns in terms of HP change as a function of the food eaten. Mathematically, the ideas of diminishing returns and saturation are well captured by the step response of a 1st order system:

Reviewer 2: ```suggestion
\subsection{Food and Health: \lstinline$updateHP$}\label{updateHP}
To increase their HP, agents need to eat. However, the amount an agent's HP improves can saturate in a single day; eating more than a certain amount will provide an agent with no extra benefit to their HP. Moreover, eating more food will lead to diminishing returns in terms of HP change. Mathematically, the ideas of diminishing returns and saturation are well captured by the step response of a 1st-order system:
```

\begin{equation}\label{updateHP_general}

>    \text{newHP}= \text{currentHP} +\underbrace{w\cdot(1-e^{-\text{foodTaken}/ \tau})}_{=\text{HPChange}}

Reviewer 2: ```suggestion
   \text{newHP}= \text{currentHP} +\underbrace{w(1-e^{\frac{-\text{foodTaken}}{\tau}})}_{\text{HPChange}}
```

\end{equation}

where the two parameters $w$ and $\tau$ are defined at the beginning of the simulation. The shape of this curve is given in \Cref{fig:updateHP} together with some important parameters.

\begin{figure}[htb]%
    \centering
    \subfloat[\centering Overview]{{\includegraphics[width=0.36\linewidth]{002_simulation_structure/images/health_updateHP_overview.pdf}}}%
    \qquad
    \subfloat[\centering Detailed representation]{{\includegraphics[width=0.36\linewidth]{002_simulation_structure/images/health_updateHP_detailed.pdf}}}%

>     \caption{\texttt{updateHP} as a function of the amount of food eaten ("FoodTaken").}%

Reviewer 2: ```suggestion
    \caption{\texttt{updateHP} as a function of the amount of food eaten (``FoodTaken'').}%
```

    \label{fig:updateHP}%
\end{figure}

> Note that it is not possible to gain more HP than $w$ over the duration of one day. Indeed, we want to prevent the agents health to change too quickly. As an example, we can think of an agent that starts from the weak level and wants to reach the maximum HP value. It would take several days for this agent to "recover" from this weak level and stabilizes its health to a large HP value. Also, it is possible for an agent to get an HP value that is larger than \texttt{maxHP} with this function. However, the obtained HP value will be reduced by the cost of living using the function \texttt{hpDecay} (see below). After this operation, the final HP value will be bounded by \texttt{maxHP}.

Reviewer 2: ```suggestion
It is not possible to gain more HP than $w$ over the duration of one day; this is an intentional limit to prevent an agent's health from improving too quickly. As an example, we can think of an agent that starts from the weak level and wants to reach the maximum HP value. It would take several days for this agent to ``recover'' from this weak level and stabilise its health to a high HP value. 

Note that it is possible for an agent to achieve an HP value that is larger than \texttt{MaxHP} inside \lstinline$hpDecay()$. At the end of each day, the \lstinline$hpDecay()$ function will apply the cost of living and then bound the final HP value by \lstinline$MaxHP$.
```

> As a special case, an agent reaching the critical state will be treated differently. For these agents, the HP update rule is the following:

Reviewer 2: ```suggestion
Agents in the critical state are treated differently. For these agents, HP is updated according to this equation:
```

\begin{equation}\label{updateHP_critical}

>     \text{newHP} = \min\left\{\text{HPCritical}+\text{HPReqCToW}, \text{currentHP} +w\cdot(1-e^{-\text{foodTaken}/ \tau})\right\}

Reviewer 2: ```suggestion
    \text{newHP} = \min\left\{\text{HPCritical}+\text{HPReqCToW}, \text{currentHP} +w(1-e^{\frac{-\text{foodTaken}}{\tau}})\right\}
```

\end{equation}

> %
> %%%%%%%% Costs of Living: hpDecay
> %
> \subsection{Costs of Living: \texttt{hpDecay}}\label{hpDecay}

Reviewer 2: ```suggestion
\subsection{Cost of Living: \lstinline$hpDecay()$}\label{hpDecay}
```

At the end of each day, the HP value of the agents will be reduced by the cost of living. The cost of living is larger for an agent with larger HP value than for an agent with lower HP value. This fact is motivated by a simple observation: humans that have stronger bodies and immune systems also need more food to sustain their level of health. The exact relation between HP value, cost of living, and HP value after applying the cost of living is given by the following linear relation:

\begin{equation}\label{hpDecay}

>     \text{newHP} = \text{currentHP}-\left[b + (\text{currentHP-WeakLevel})\cdot s\right]

Reviewer 2: ```suggestion
    \text{newHP} = \text{currentHP}-\left[b + s(\text{currentHP}-\text{WeakLevel})\right]
```

\end{equation}

> where $b$ is a (constant) base cost, and $s$ is the slope of the linear function. These parameters are initialized at the beginning of the simulation. \ToDo{Add a diagram for this?}

Reviewer 2: ```suggestion
where $b$ is a (constant) base cost, and $s$ is the slope of the linear function. These parameters are initialised at the beginning of the simulation. \ToDo{Add a diagram for this?}
```

> To ensure that the HP value at the end of the day is bounded by \texttt{maxHP}, we slightly modify the above expression:

Reviewer 2: ```suggestion
To ensure that the HP value at the end of the day is bounded by \texttt{MaxHP}, we slightly modify the above expression:
```

\begin{equation}\label{hpDecay_bounded}

>     \text{newHP} =\max\left\{\text{MaxHP}, \text{currentHP}-\left[b + (\text{currentHP-WeakLevel})\cdot s\right]\right\}

Reviewer 2: ```suggestion
    \text{newHP} =\max\left\{\text{MaxHP}, \text{currentHP}-\left[b + s(\text{currentHP}-\text{WeakLevel})\right]\right\}
```

\end{equation}

> Once again, the agents in the critical state are treated differently. If, by the end of the day, they achieved to gain \texttt{HPReqCToW} HP, meaning if their HP after eating is

Reviewer 2: ```suggestion
For agents in the critical state that gain \texttt{HPReqCToW} HP in a single day, i.e. their HP after eating is
```

\begin{equation}\label{HPReqCToW}

>     \text{currentHP} \geq \text{HPCritical}+\text{HPReqCToW}

Reviewer 2: ```suggestion
    \text{currentHP} \geq \text{HPCritical}+\text{HPReqCToW},
```

\end{equation}

> they will be upgraded to the \texttt{WeakLevel}:

Reviewer 2: ```suggestion
their HP will be set to \texttt{WeakLevel}:
```

\begin{equation}\label{hpDecay_critical_upgrade}
    \text{newHP} = \text{WeakLevel}
\end{equation}

> On the other hand, if they do not fulfill the above condition, They will be kept in the critical state by setting their HP to

Reviewer 2: ```suggestion
Agents in the critical state which do not manage to improve their HP by \lstinline$HPReqCToW$ will be kept in the critical state:
```

\begin{equation}\label{hpDecay_critical_stay}
    \text{newHP} = \text{HPCritical}
\end{equation}

> and a counter named \texttt{daysAtCritical} will be increased by one. If \texttt{daysAtCritical} reaches \texttt{MaxDayCritical}, the agents die and are replaced. Note that the counter is reset to 0 if the agent exit the critical state.

Reviewer 2: ```suggestion
with the \texttt{daysAtCritical} counter incremented by 1. If \texttt{daysAtCritical} reaches \texttt{MaxDayCritical}, the agent dies and is replaced. This counter is reset to 0 if an agent exits the critical state.
```
